### ISSUE TITLE
Adding RECO era to process history

### ISSUE THREAD

**Issue by osschar** [cms-sw/cmssw#29092]
For Fireworks to be able to reliably identify which RECO geometry to use for a given EDM file we propose to add RECO era string to process history and thus to all EDM files produced in the future. 

This string can then be mapped to RECO geometry version that should be used by Fireworks to cover 95%+ of users use-cases (special cases should use full-framework version or generate their own reco geometry).

Apparently the RECO era exists only as a python variable and would have to be added to a new or an existing tracked pset.

**Comment by Dr15Jones** [comment]
assign core

**Comment by makortel** [comment]
assign core, visualization, reconstruction

**Comment by makortel** [comment]
@osschar Could you describe if there are any other possible sources for this information, and what their drawbacks would be? How are old files (that would not have this information) be dealt with? (and why that would not work for the future)?

**Comment by makortel** [comment]
I believe we have the "era" as a string only in `cmsDriver`/`ConfigBuilder`. The `cms.Process` gets only a list of `cms.Modifier` objects and has little-to-no means to find out their "names".

**Comment by slava77** [comment]
> I believe we have the "era" as a string only in `cmsDriver`/`ConfigBuilder`. The `cms.Process` gets only a list of `cms.Modifier` objects and has little-to-no means to find out their "names".

is there a way to inject the name of at least the ModifierChain "Era" during the cms.Process construction?
If the individual modifiers can enter as well, perhaps this can be also useful in resolution of #13326

**Comment by makortel** [comment]
> is there a way to inject the name of at least the ModifierChain "Era" during the cms.Process construction?

Not at the moment. And because the era name would be essentially meaningless to the `cms.Process`,  I'd think other means would be simpler to get such string an a tracked `PSet`.

> If the individual modifiers can enter as well, perhaps this can be also useful in resolution of #13326

I don't see how that would relate to cfi generation from `fillDescriptions()`.

**Comment by fwyzard** [comment]
Isn't the geometry something that should be tracked by the global tag, rather then the era ?

**Comment by ianna** [comment]
> Isn't the geometry something that should be tracked by the global tag, rather then the era ?

yes, and it's the case for all cms jobs that use it from a GT. It's not the case for the event display and neither for the 2026 workflows.

**Comment by fwyzard** [comment]
> It's not the case for the event display ... 

Sure - but that means that the even display could extrack the global tag used in the RECO jobs from the provenance, and use that to set up its geometry.

> ... and neither for the 2026 workflows.

Well, that is simply something that should be fixed :-)
(we were supposed to have modular global tags years ago...)

**Comment by osschar** [comment]
> @osschar Could you describe if there are any other possible sources for this information, and what their drawbacks would be? How are old files (that would not have this information) be dealt with? (and why that would not work for the future)?

Currently, we don't do any checks, default reco geometry is defined per release as the reco geometry currently used for data. Users have to specify a non-default fireworks reco geometry file on the command line.

As for possible sources, there is:
- run number  - but this does not work for MC; and
-  GT which changes too frequently and does not have a good naming convention. 
Using this, we will have a long, error prone, and very likely incomplete section of ifs and regexp matches in cmsShow code. This gets especially problematic for HL-LHC productions --- and we are seeing people walking into this minefiled --- and this is also the main motivation for this request.

For old files we still plan to use run number for data and GT for MC.

We plan to add the same "reco era" tag to fireworks reco geometry files. The main purpose is to detect that something is wrong before we silently show a wrong information without even knowing it.
Secondary, we can load the proper reco geometry based on reco era information in the EDM file.

**Comment by makortel** [comment]
@Dr15Jones and I had a short discussion. In principle one possibility could be to add an option for `cms.ModifierChain` (that all the "eras" in https://github.com/cms-sw/cmssw/tree/master/Configuration/Eras/python are) to have a label as well (implying that the "era name" would need to be specified twice), and then make `cms.Process` to fill some tracked `cms.vstring()` in some tracked `PSet` based on the values the `cms.ModifierChain`s the `Process` got in its constructor. This solution would be more general than patching `cmsDriver`/`ConfigBuilder`.

Then the consumer side would have to parse the "era"
* no era -> run 1
* "Run1" in era -> run1
* "Run2" in era without a year -> run2 2015
* "Run2" in era with a year -> run2 of that year
* "Run3" in era -> run3
* "Phase2" in era -> phase2, but the logic actually breaks down because we don't have a different era for each geometry variation (apparently only for different calorimeter versions)

**Comment by slava77** [comment]
> "Phase2" in era -> phase2, but the logic actually breaks down because we don't have a different era for each geometry variation (apparently only for different calorimeter versions)

This is a good point!
It means that saving an era is not enough to give an automatic config/geometry "barcode" needed for the event display.

This issue perhaps needs to be extended with the request to save the argument in the  --geometry  selector in the cmsDriver commands (as in ```--geometry Extended2026D49).

**Comment by makortel** [comment]
Should user-produced output (using non-cmsDriver-generated, or -generated-but-edited configuration files) be supported as well?

**Comment by davidlange6** [comment]
I believe that --geometry values  are only relevant until geometries get into the gt.

On Mar 4, 2020 4:57 PM, Slava Krutelyov <notifications@github.com> wrote:

"Phase2" in era -> phase2, but the logic actually breaks down because we don't have a different era for each geometry variation (apparently only for different calorimeter versions)

This is a good point!
It means that saving an era is not enough to give an automatic config/geometry "barcode" needed for the event display.

This issue perhaps needs to be extended with the request to save the argument in the --geometry selector in the cmsDriver commands (as in ```--geometry Extended2026D49).

**Comment by slava77** [comment]
> I believe that --geometry values are only relevant until geometries get into the gt.

well, this is not practical, since I guess that the phase-2 GT geometry will not show up until 2028.

@ianna is there a string in the XML geom builder config (provenance) that could serve as an appropriate key to discover the geometry kind?

**Comment by davidlange6** [comment]
On Mar 4, 2020, at 5:12 PM, Slava Krutelyov <notifications@github.com<mailto:notifications@github.com>> wrote:


I believe that --geometry values are only relevant until geometries get into the gt.

well, this is not practical, since I guess that the phase-2 GT geometry will not show up until 2028.


Would be a pity - Due to what constraint?

**Comment by slava77** [comment]
> well, this is not practical, since I guess that the phase-2 GT geometry will not show up until 2028. Would be a pity - Due to what constraint?

perhaps @ianna and @kpedro88 could clarify that my guess is wrong

**Comment by kpedro88** [comment]
Phase 2 geometries are constantly in development and fluctuating. So far, the idea to put these geometries into global tags has been seen as operational overhead without any benefit.

In general (without having followed this entire conversation), I would propose some method to store the provenance of the geometry generically, whether from XML or GT.

**Comment by fwyzard** [comment]
> Phase 2 geometries are constantly in development and fluctuating. So far, the idea to put these geometries into global tags has been seen as operational overhead without any benefit.

The main benefit would be that all the people that use the MC samples would not need to decode the arcane incantations used to set up the geometry, but could simply use the global tag, like we have been doing for Run 1 and Run 2 for the past 10 years.

At some point Phase 2 needs to stop being a playground.

**Comment by kpedro88** [comment]
runTheMatrix.py is very easy to use, and ensures the correct eras, geometries, etc. are kept together.

I don't really want to encourage users to try to put together their own configs ex nihilo, because that generates a lot of annoying emails with trivial mistakes that would have been avoided by using approved methods.

**Comment by fwyzard** [comment]
> runTheMatrix.py is very easy to use, and ensures the correct eras, geometries, etc. are kept together.

runTheMatrix.py covers the central, offline workflows.
It does not cover the online workflows (i.e. the HLT and online DQM) nor the user analyses.

So, it's a very useful tool for centrally maintained workflows, because it hides a lot of the complexities that make the life miserable for the people that work on other topics.

In an ideal world the offline reconstruction developers, conveners and release managers would stop adding one layer of customisation on top of an other.

Sticking to the example of the geometry, it can be set by
  - the configuration
  - the global tags
  - an @auto global tags
  - a customisations that add arbitrary changes to a configuration and/or a global tag
  - an era
  - a customisation on top of an era

So, sure, why not create an additional field that will supposedly store the actual geometry being used, at least until somebody comes up with an other idea to "simplify" the developments...

**Comment by kpedro88** [comment]
Geometries are never set by Eras. That idea was discussed early on for phase 2 but rejected by the geometry team as too error-prone.

There are mechanisms to add special workflows to the matrix, so it could easily be extended to cover online workflows etc.

**Comment by davidlange6** [comment]
Sure. Presumably this changes on the timescale of a few years or less as designs are settled?

On Mar 4, 2020 5:51 PM, Kevin Pedro <notifications@github.com> wrote:

Phase 2 geometries are constantly in development and fluctuating. So far, the idea to put these geometries into global tags has been seen as operational overhead without any benefit.

In general (without having followed this entire conversation), I would propose some method to store the provenance of the geometry generically, whether from XML or GT.

**Comment by kpedro88** [comment]
Indeed, I have fond hopes that the designs will be settled before we start actually building them.

**Comment by fwyzard** [comment]
> There are mechanisms to add special workflows to the matrix, so it could easily be extended to cover online workflows etc.

Are you suggesting that we should extend runTheMatrix to support workflows that are not based on cmsDriver ?

**Comment by ianna** [comment]
The geometry version is uniquely identified by a GT and is recorded in Provenance, but I do not see any suitable place where and how to add a version without adding an extra burden on its maintenance.

For example, here we know an exact geometry version:
```python
ESSource: GlobalTag SIM
 parameters: {
  @module_edm_type: string tracked  = 'ESSource'
  @module_label: string tracked  = 'GlobalTag'
  @module_type: string tracked  = 'PoolDBESSource'
  connect: string tracked  = 'frontier://FrontierProd/CMS_CONDITIONS'
  globaltag: string tracked  = '110X_mcRun1_realistic_v2'
  snapshotTime: string tracked  = ''
  DBParameters: PSet tracked = ({
  })
  toGet: VPSet tracked = ({
  })
 }
...
ESModule: XMLFromDBSource SIM
 parameters: {
  @module_edm_type: string tracked  = 'ESProducer'
  @module_label: string tracked  = 'XMLFromDBSource'
  @module_type: string tracked  = 'XMLIdealGeometryESProducer'
  appendToDataLabel: string tracked  = ''
  label: string tracked  = 'Extended'
  rootDDName: string tracked  = 'cms:OCMS'
 }
```
On the other hand, this is what recorded for the 2026 scenarios and it's pretty much impossible to extract an exact geometry version. One would need an exact CMSSW release to do so. 
```python
ESSource: XMLIdealGeometryESSource SIM
 parameters: {
  @module_edm_type: string tracked  = 'ESSource'
  @module_label: string tracked  = ''
  @module_type: string tracked  = 'XMLIdealGeometryESSource'
  geomXMLFiles: vstring tracked  = {'Geometry/CMSCommonData/data/materials.xml','Geometry/CMSCommonData/data/rotations.xml','Geometry/CMSCommonData/data/extend/v2/cmsextent.xml','Geometry/CMSCommonData/data/cms/2026/v2/cms.xml','Geometry/CMSCommonData/data/eta3/etaMax.xml','Geometry/CMSCommonData/data/cmsMother.xml','Geometry/CMSCommonData/data/cmsTracker.xml','Geometry/CMSCommonData/data/caloBase/2026/v2/caloBase.xml','Geometry/CMSCommonData/data/cmsCalo.xml','Geometry/CMSCommonData/data/muonBase/2026/v2/muonBase.xml','Geometry/CMSCommonData/data/cmsMuon.xml','Geometry/CMSCommonData/data/mgnt.xml','Geometry/CMSCommonData/data/beampipe/2026/v1/beampipe.xml','Geometry/CMSCommonData/data/cmsBeam/2026/v1/cmsBeam.xml','Geometry/CMSCommonData/data/muonMB.xml','Geometry/CMSCommonData/data/muonMagnet.xml','Geometry/CMSCommonData/data/cavern/2017/v2/cavern.xml','Geometry/CMSCommonData/data/cavernData/2017/v1/cavernData.xml','Geometry/CMSCommonData/data/cavernFloor/2017/v1/cavernFloor.xml','Geometry/TrackerCommonData/data/PhaseII/trackerParameters.xml','Geometry/TrackerCommonData/data/pixfwdCommon.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker613/pixfwd.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker404/pixbar.xml','Geometry/TrackerCommonData/data/trackermaterial.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker404/otst.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker613/tracker.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker613/pixel.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker404/trackerbar.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker404/trackerfwd.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker404/trackerStructureTopology.xml','Geometry/TrackerCommonData/data/PhaseII/TiltedTracker613/pixelStructureTopology.xml','Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/trackersens.xml','Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/pixelsens.xml','Geometry/TrackerRecoData/data/PhaseII/TiltedTracker613/trackerRecoMaterial.xml','Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/trackerProdCuts.xml','Geometry/TrackerSimData/data/PhaseII/TiltedTracker404/pixelProdCuts.xml','Geometry/TrackerSimData/data/trackerProdCutsBEAM.xml','Geometry/EcalCommonData/data/eregalgo/2026/v2/eregalgo.xml','Geometry/EcalCommonData/data/ectkcable/2026/v1/ectkcable.xml','Geometry/EcalCommonData/data/ectkcablemat/2026/v1/ectkcablemat.xml','Geometry/EcalCommonData/data/ebalgo.xml','Geometry/EcalCommonData/data/ebcon.xml','Geometry/EcalCommonData/data/ebrot.xml','Geometry/HcalCommonData/data/hcalrotations.xml','Geometry/HcalCommonData/data/hcal/v2/hcalalgo.xml','Geometry/HcalCommonData/data/hcalbarrelalgo.xml','Geometry/HcalCommonData/data/hcalcablealgo/v2/hcalcablealgo.xml','Geometry/HcalCommonData/data/hcalouteralgo.xml','Geometry/HcalCommonData/data/hcalforwardalgo.xml','Geometry/HcalCommonData/data/hcalSimNumbering/NoHE/hcalSimNumbering.xml','Geometry/HcalCommonData/data/hcalRecNumbering/NoHE/hcalRecNumbering.xml','Geometry/HcalCommonData/data/average/hcalforwardmaterial.xml','Geometry/HGCalCommonData/data/hgcalMaterial/v1/hgcalMaterial.xml','Geometry/HGCalCommonData/data/hgcal/v10/hgcal.xml','Geometry/HGCalCommonData/data/hgcalEE/v10/hgcalEE.xml','Geometry/HGCalCommonData/data/hgcalHEsil/v10/hgcalHEsil.xml','Geometry/HGCalCommonData/data/hgcalHEmix/v10/hgcalHEmix.xml','Geometry/HGCalCommonData/data/hgcalwafer/v9/hgcalwafer.xml','Geometry/HGCalCommonData/data/hgcalcell/v9/hgcalcell.xml','Geometry/HGCalCommonData/data/hgcalCons/v10/hgcalCons.xml','Geometry/MuonCommonData/data/mbCommon/2017/v2/mbCommon.xml','Geometry/MuonCommonData/data/mb1/2015/v1/mb1.xml','Geometry/MuonCommonData/data/mb2/2015/v1/mb2.xml','Geometry/MuonCommonData/data/mb3/2015/v1/mb3.xml','Geometry/MuonCommonData/data/mb4/2015/v1/mb4.xml','Geometry/MuonCommonData/data/muonYoke/2021/v2/muonYoke.xml','Geometry/MuonCommonData/data/mf/2026/v2/mf.xml','Geometry/MuonCommonData/data/rpcf/2026/v1/rpcf.xml','Geometry/MuonCommonData/data/gemf/TDR_BaseLine/gemf.xml','Geometry/MuonCommonData/data/gem11/TDR_BaseLine/gem11.xml','Geometry/MuonCommonData/data/gem21/TDR_Dev/gem21.xml','Geometry/MuonCommonData/data/csc/2015/v1/csc.xml','Geometry/MuonCommonData/data/mfshield/2026/v1/mfshield.xml','Geometry/MuonCommonData/data/me0/TDR_Dev/me0.xml','Geometry/ForwardCommonData/data/forwardshield/2017/v1/forwardshield.xml','Geometry/ForwardCommonData/data/brmrotations.xml','Geometry/ForwardCommonData/data/PostLS2/brm.xml','Geometry/ForwardCommonData/data/zdcmaterials.xml','Geometry/ForwardCommonData/data/lumimaterials.xml','Geometry/ForwardCommonData/data/zdcrotations.xml','Geometry/ForwardCommonData/data/lumirotations.xml','Geometry/ForwardCommonData/data/zdc.xml','Geometry/ForwardCommonData/data/zdclumi.xml','Geometry/ForwardCommonData/data/cmszdc.xml','Geometry/MTDCommonData/data/btl.xml','Geometry/MTDCommonData/data/etl/v2/etl.xml','Geometry/MTDCommonData/data/CrystalBarPhiFlat/v2/mtd.xml','Geometry/MTDCommonData/data/CrystalBarPhiFlat/mtdStructureTopology.xml','Geometry/MTDCommonData/data/CrystalBarPhiFlat/mtdParameters.xml','Geometry/MuonCommonData/data/muonNumbering/TDR_DeV/muonNumbering.xml','Geometry/EcalSimData/data/PhaseII/ecalsens.xml','Geometry/HcalCommonData/data/hcalsens/NoHE/hcalsenspmf.xml','Geometry/HcalSimData/data/hf.xml','Geometry/HcalSimData/data/hfpmt.xml','Geometry/HcalSimData/data/hffibrebundle.xml','Geometry/HcalSimData/data/CaloUtil.xml','Geometry/HGCalSimData/data/hgcsensv9.xml','Geometry/MuonSimData/data/PhaseII/ME0EtaPart/muonSens.xml','Geometry/DTGeometryBuilder/data/dtSpecsFilter.xml','Geometry/CSCGeometryBuilder/data/cscSpecsFilter.xml','Geometry/CSCGeometryBuilder/data/cscSpecs.xml','Geometry/RPCGeometryBuilder/data/2026/v1/RPCSpecs.xml','Geometry/GEMGeometryBuilder/data/v7/GEMSpecsFilter.xml','Geometry/GEMGeometryBuilder/data/v7/GEMSpecs.xml','Geometry/ForwardCommonData/data/brmsens.xml','Geometry/ForwardSimData/data/zdcsens.xml','Geometry/MTDSimData/data/CrystalBarPhiFlat/mtdsens.xml','Geometry/HcalSimData/data/HcalProdCuts.xml','Geometry/EcalSimData/data/EcalProdCuts.xml','Geometry/HGCalSimData/data/hgcProdCutsv9.xml','Geometry/MuonSimData/data/PhaseII/muonProdCuts.xml','Geometry/ForwardSimData/data/zdcProdCuts.xml','Geometry/ForwardSimData/data/ForwardShieldProdCuts.xml','Geometry/MTDSimData/data/CrystalBarPhiFlat/mtdProdCuts.xml','Geometry/CMSCommonData/data/FieldParameters.xml'}
  rootNodeName: string tracked  = 'cms:OCMS'
 }

```

> > well, this is not practical, since I guess that the phase-2 GT geometry will not show up until 2028. Would be a pity - Due to what constraint?
> 
> perhaps @ianna and @kpedro88 could clarify that my guess is wrong

There is no constraint, just a policy and lack of manpower. 

Unfortunately, this is already hurting us, because we cannot guarantee the same geometry version at the RECO step running separately from the SIM step. Here is a quote from @tomalin (see full conversation [in Geometry HN forum](https://hypernews.cern.ch/HyperNews/CMS/get/geometry/298.html)):

1) MC production for the upgrade L1 trigger studies is a two step process. The first makes GEN-SIM-DIGI-RAW datasets specifying a certain CMS geometry (e.g. D21), & the second creates the L1 trigger objects specifying which reco geometry to use (e.g. D21Reco). The geometries specified for these two steps must be consistent with each other (e.g. Both D21). However, if we make a mistake by specifying two inconsistent geometries (D21 + D41Reco), then CMSSW doesn’t always crash, but can just produces slightly incorrect L1 results. This is dangerous. How can we force a fatal error to be thrown?

2) The cabling map indicating which Tracker modules are connected to which off-detector readout electronics has been recently added to the CMSSW DB. This map is different for each new CMS geometry. So if we are use D21 geometry, we need to ensure that our CMSSW L1 track jobs picks up the correct map from the DB. But since our jobs just specify to use the “auto:upgradePLS3” GlobalTag, I fear that this may not be happening.

**Comment by ianna** [comment]
Would it be possible to assign a different (other then 1) run number to SIM? One could think of something like:
  * 2015 geometry == SIM Run 1 to Run 10
  * 2021 geometry == SIM Run 10 to Run 20
  * 2026DX geometry == SIM Run X (based on a geometry version)

**Comment by ianna** [comment]
An option **to version SIM geometry** could be to add an auto-generated version directly in an auto-generated geometry scenario. 

To do that:
  * Firstly, we'd need to make sure that all geometry scenarios are auto-generated (similar to [2026](https://cmssdt.cern.ch/lxr/source/Configuration/Geometry/python/dict2026Geometry.py). 
  * Secondly, the script which generates the scenarios should generate a `version.xml` with an appropriate XML tag and attribute and include it in the scenario.
 * Thirdly, the DD code and explicitly the XML parser should be modified to process the XML tag/attribute. Then, all the payloads should be updated to include the geometry version.

Note, this does not solve **RECO geometry versioning** which is uniquely identified by a GT and not all GTs contain SIM geometry.

**Comment by makortel** [comment]
> we cannot guarantee the same geometry version at the RECO step running separately from the SIM step

This is a more general issue than just geometry: we are not able to check in a later step if the conditions are compatible with the conditions of an earlier step. For example, during 2015-2017 I recall the SIM and RECO beamspot becoming inconsistent about once a year or so. Fortunately that was relatively easy to spot because tracking performance went haywire.

**Comment by kpedro88** [comment]
@fwyzard 
> Are you suggesting that we should extend runTheMatrix to support workflows that are not based on cmsDriver ?

Maybe it makes more sense for the existing HLT infrastructure to import phase 2 workflow properties directly from [upgradeWorkflowComponents.py](https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py) to ensure consistency.

**Comment by kpedro88** [comment]
@ianna

> However, if we make a mistake by specifying two inconsistent geometries (D21 + D41Reco), then CMSSW doesn’t always crash, but can just produces slightly incorrect L1 results.

I agree it would be better to enforce this automatically, but in general these mistakes happen because people aren't following approved procedures, i.e. runTheMatrix.

> Would it be possible to assign a different (other then 1) run number to SIM?

There is code scattered throughout CMSSW (as well as user repositories) that assumes SIM has run = 1, so I don't think this is a good idea.

**Comment by ianna** [comment]
> > we cannot guarantee the same geometry version at the RECO step running separately from the SIM step
> 
> This is a more general issue than just geometry: we are not able to check in a later step if the conditions are compatible with the conditions of an earlier step. For example, during 2015-2017 I recall the SIM and RECO beamspots becoming inconsistent about once a year or so. Fortunately that was relatively easy to spot because tracking performance went haywire.

Well, yes. It's much simpler for the SIM and RECO geometries though. The versions are identical if the jobs are using the same GT line.

**Comment by kpedro88** [comment]
also

> So if we are use D21 geometry, we need to ensure that our CMSSW L1 track jobs picks up the correct map from the DB.

Tracker geometry-dependent records are specified in [autoCondPhase2.py](https://github.com/cms-sw/cmssw/blob/master/Configuration/AlCa/python/autoCondPhase2.py) and the proper GT for each workflow is specified in [upgradeWorkflowComponents.py](https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py). So... the answer is to use runTheMatrix. (Do I sound like a broken record yet?)

**Comment by fwyzard** [comment]
> @fwyzard
> 
> > Are you suggesting that we should extend runTheMatrix to support workflows that are not based on cmsDriver ?
> 
> Maybe it makes more sense for the existing HLT infrastructure to import phase 2 workflow properties directly from [upgradeWorkflowComponents.py](https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py) to ensure consistency.

> So... the answer is to use runTheMatrix. (Do I sound like a broken record yet?)

Yes.

Honestly, it would make sense for people to stop assuming that the offline reconstruction is the only thing CMSSW should cater for.

Or at least to stop commenting about things they know very little about ?

**Comment by makortel** [comment]
> > Would it be possible to assign a different (other then 1) run number to SIM?
> 
> There is code scattered throughout CMSSW (as well as user repositories) that assumes SIM has run = 1, so I don't think this is a good idea.

These assumptions need to be removed if/when we get to run-dependent MC (that *will* contain non-1 run numbers).

**Comment by kpedro88** [comment]
runTheMatrix workflows are tested daily in PRs and IBs, ensuring that they do not break and provide consistent results. Non-offline workflows would be well served to imitate this successful strategy.

**Comment by fwyzard** [comment]
It's amazing, you keep commenting and showing that you have absolutely no idea what you are taling about.

**Comment by fwyzard** [comment]
@kpedro88, basically what you are suggesting is "I developed a particular solution for a very specific problem, so every body else should simply use it", even if the problem they need to solve is fundamentally different.

In principle, yes, the HLT configuration could move back into the release.

However, to avoid issues sneaking in the HLT configurations, that would require the HLT L2s to have veto/approval powers on every single reconstruction, configuration, and conditions package (which would not be a bad thing now that I think about) in the same way as the RECO, Operations, AlCa, etc. conveners do.

It would also require CMS to build a new release every time an update to the HLT configuration is needed - which, during data taking, means roughly once or twice a week.

Assuming this is something that O&C is not ready to accept, we need to stick with the current, less-than-ideal solution, where the HLT configuration is maintained outside of the main CMSSW release and is dumped in there on a semi-regular base.

In fact, for the Phase 2 developments we are trying a hybrid approach, where the configuration lives in a separate git repository (rather than in oracle) and is based on individual python files (rather than a monolithic dump); the aim is to re-use as much as possible of the individual components from the offline configurations (i.e. cfi files from the release or autogenerated from fillDescriptions()), and only maintain the HLT-specific configurations.

However, this still requires building self-contained configurations that do not use `cmsDriver` - and by extension `runTheMatrix`.

As a side note: the HLT configurations are also tested regularly on IBs (look them up in the addon tests).

**Comment by fwyzard** [comment]
Now, since this is clearly going way off topic, I'll be happy to cut the discussion here.

Feel free to open a separate issue if you have concrete proposals about improving the situation.

**Comment by makortel** [comment]
Let me try to summarize the options presented so far (for MC)

* configuration era
   * could be used to figure out run1-3 geometries, not sufficient to figure out the used phase2 geometry (less eras than geometries)
   * would need some additional work to store it in the provenance
* global tag name
   * could be used to figure out run1-3 geometries, not sufficient to figure out the used phase2 geometry (currently one GT for all phase2 workflows, phase2 geometry comes from XML file)
* XML geometry ESProducer configuraton
   * figuring out the phase2 geometry from the list of XML files is nontrivial, in general depends on CMSSW version
* make cmsDriver to store values of `--era`, `--conditions`, and `--geometry` in tracked strings so they can be read from the provenance
   * does not help non-cmsDriver workflows (even though the same strings could be filled manually)
* come up with some new mechanism to version the "detector configuration" and enforce compatibility between input data and the job configuration (wild idea)
* use run numbers in MC
   * `run=1` assumed here and there

In general the provenance system was not designed to be used to configure any later processing step (including visualization). For GT and XML geometry one should also check any ESPrefers in that configuration.

The run numbers do define the detector configuration uniquely for data, in a sense it would be natural to use them for the same purpose for MC as well.
